Implement "miss cache" for TTreeCache

bbockelm opened the pull request:
The _miss cache_, implemented in this pull request, implements an optimization when the TTreeCache fails to work.

The miss cache will keep track of any branch that has been accessed; when there is a TTC miss, it automatically fetches the current basket for all active branches.  This should have a worst case read size equal to the size of the file's cluster size, but potentially a significant savings in the number of IO operations.  The latter is extremely useful if we're doing IO on high-latency links.

This optimization works well for the "trigger pattern," where the user may examine a number of branches and, when the event contents for those branches passes a particular filter, reads out the remaining branches.  If there are 100 additional branches, this would do all reads in a single network round-trip as opposed to 100 round trips.

The approach has served us well in CMS and been utilized as a layer on top of ROOT for about 3 years.

Unfortunately, we must iterate through a set of branches and find the correct basket.  This is not necessarily a cheap CPU operation and may be too expensive if the underlying filesystem is SSD-based.  Hence, we turn this optimization off by default.

bbockelm commented:
@pcanal - turned this into a formal PR and tried to fix up as many code comments you suggested previously.  Probably a good point to re-review.

Or go on vacation.  Whichever is your preference.  :)

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -77,10 +129,14 @@ class TTreeCache : public TFileCacheRead {
    virtual Int_t        DropBranch(const char *branch, Bool_t subbranches = kFALSE);
    virtual void         Disable() {fEnabled = kFALSE;}
    virtual void         Enable() {fEnabled = kTRUE;}
+   void                 SetOptimizeMisses(bool opt);
```
In this file/section the routine are (and thus should stay) listed alphabetically.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -57,6 +57,17 @@ The learning period is stopped (and prefetching is actually started) when:
+but the majority of event collections are read only when the trigger results
+pass a set of filters.  NOTE - when this mode is enabled, the memory dedicated
+to the cache will up to double in the case of cache miss.
+
```
The PR blur states "Unfortunately, we must iterate through a set of branches and find the correct basket. This is not necessarily a cheap CPU operation and may be too expensive if the underlying filesystem is SSD-based. Hence, we turn this optimization off by default."   

Some equivalent wording ought to be in this description.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -240,8 +251,22 @@ of effective system reads for a given file with a code like
+#include "TMath.h"
 #include <limits.h>
 
+// TODO: Copied from TBranch.cxx
```
This is now in 3 files (also in TBranchElement).  We need to move it to a single location (maybe RConfig.h)

bbockelm replied:
:) Even the TODO is in three files now.

Is `RConfig.h` part of the public API?  If so, do we want to put this into the public API?

I'll prepare a separate PR.

bbockelm replied:
This is now in #242.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -613,11 +638,253 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+}
+
+
+TBranch *TTreeCache::CalculateMissEntries(Long64_t pos_in, int len_in, bool all) {
```
All routines need to have a doxygen docs in front of the function.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -613,11 +638,253 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+//// Start of methods for the miss cache.
+///////////////////////////////////////////////////////////////////////////////
+
+/**
```
The doc start pattern for functions should be
```
////////////////////////////////////////////////////////////////////////////////
///
``` rather than C style and the content should be useable by doxygen.

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -62,10 +70,54 @@ class TTreeCache : public TFileCacheRead {
 
+   // These members hold cached data for missed branches when miss optimization
+   // is enabled.  Pointers are only initialized if the miss cache is enabled.
+   bool            fOptimizeMisses {false}; //! true if we should optimize cache misses.
```
For consistency use Bool_t

pcanal replied:
For consistency use Bool_t and kFALSE

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -62,10 +70,54 @@ class TTreeCache : public TFileCacheRead {
+      std::vector<TBranch*> fBranches;  //! list of branches that we read on misses.
+      std::vector<char> fData;  //! Actual data in the cache.
+
+      void clear() {fEntries.clear(); fBranches.clear(); fData.clear();}
```
Please add white spaces : { fEntries .. clear(); }

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -77,6 +129,7 @@ class TTreeCache : public TFileCacheRead {
    virtual Int_t        DropBranch(const char *branch, Bool_t subbranches = kFALSE);
    virtual void         Disable() {fEnabled = kFALSE;}
    virtual void         Enable() {fEnabled = kTRUE;}
+   bool                 GetOptimizeMisses() const {return fOptimizeMisses;}
```
Please add white spaces { re....; }

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -85,6 +138,8 @@ class TTreeCache : public TFileCacheRead {
    virtual Int_t        GetEntryMax() const {return fEntryMax;}
    static Int_t         GetLearnEntries();
    virtual EPrefillType GetLearnPrefill() const {return fPrefillType;}
+   double               GetMissEfficiency() const;
```
For consistency, please use Double_t

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+/// The first time this is called on a TTreeCache object, the corresponding
+/// data structures will be allocated.  Subsequent enable / disables will
+/// simply turn the functionality on/off.
+void TTreeCache::SetOptimizeMisses(bool opt) {
```
Bool_t

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+///   this IO operation.
+/// - If no corresponding branch could be found (or an error occurs), this
+///   returns nullptr.
+TBranch *TTreeCache::CalculateMissEntries(Long64_t pos_in, int len_in, bool all) {
```
Int_t and Bool_t

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+   fMissCache->fData.reserve(cumulative);
+   //printf("Reading %lu bytes into miss cache for %lu entries.\n", cumulative, fEntries->size());
+   fNMissReadPref += fMissCache->fEntries.size();
+   fFile->ReadBuffers(&(fMissCache->fData[0]), &positions[0], &lengths[0], fMissCache->fEntries.size());
```
For consistency always use the pattern &(...[0])  i.e. add missing paranthesis

pcanal commented:
Except for minor coding conventions update, this looks ready to go and will be added to the master after the branching of the v6.08 release.   It would also be best to extent roottest to include a test of this functionality.

bbockelm commented:
Ok, I think the code formatting issues mentioned previously are now all resolved.

I have a unit test ready, but I was going to wait for the roottest PR until this one is merged.

pcanal commented:
> I have a unit test ready, but I was going to wait for the roottest PR until this one is merged.

I tend to upload both the test and the patch at roughly the same time (allows to check whether the patch is indeed working :) in addition to not break other stuff)

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+   auto pos = static_cast<ULong64_t>(pos_in);
+   auto len = static_cast<UInt_t>(len_in);
+
+   int count = all ? (fTree->GetListOfLeaves())->GetEntriesFast() : fMissCache->fBranches.size();
```
Would itn't make more sense to instead have a pointer to the 'active' collection instead (thus avoid the second if ?

bbockelm replied:
Not clear to me there's an easy way to do this:

  - If `all == false`, then we are iterating across a `std::vector<Branch*>`.
  - If `all == true`, then we are iterating over a `TObjArray` of `TLeaf*`.

Hence, one needs a little bit of logic per loop iteration to get to the correct object.

pcanal replied:
Fair enough, I missed that difference.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+/// Returns:
+/// - IOPos describing the IO operation necessary for the basket on this branch
+/// - On failure, IOPos.length will be set to 0.
+TTreeCache::IOPos TTreeCache::FindBranchBasket(TBranch &b) {
```
As this does not return a TBasket, a better (more consistent) name would be FindBranchBasketPos.

Also, I found (reading the code) that it would have helped me to make the entry number to be a explicit parameter (instead of being an implicit one on line 700).

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+      return IOPos{0, 0};
+   }
+
+   return IOPos{static_cast<ULong64_t>(pos), static_cast<UInt_t>(len)};
```
Why elect to have the member of IOPos being unsigned rather than being the 'natural' type (and thus avoid the static cast while not losing anything since the interface takes/uses the 'natural' type)?

bbockelm replied:
I think this is just inspired from how it is done in CMS.  I can switch it back to `Long64_t` to make it more "ROOT-like".

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+   ULong64_t cumulative = 0;
+   for (auto &mcentry : fMissCache->fEntries) {
+         positions.push_back(mcentry.fIO.fPos);
+         lengths.push_back(mcentry.fIO.fLen);
```
Shouldn't the indentation be 3 characters?

bbockelm replied:
Yup.  I assume that's due to the fact this PR predates `clang-format` checks; as soon as I update, Travis-CI should spit out such fixes.

bbockelm commented:
@pcanal - I rebased this on master, hopefully making the tests and such more relevant.  I believe I addressed your review comments.

Seems to have also kicked off the CI tests successfully; I'll likely have a few `clang-format` items to fix up from those.

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -33,35 +38,84 @@ class TTreeCache : public TFileCacheRead {
+   std::unique_ptr<MissCache> fMissCache;  //! Cache contents for misses
 
 private:
    TTreeCache(const TTreeCache &);            //this class cannot be copied
```
while we are at it.  Can you replace this patterns with = delete;

pcanal commented on tree/tree/inc/TTreeCache.h:
```diff
@@ -33,35 +38,84 @@ class TTreeCache : public TFileCacheRead {
+
+      Long64_t fPos{0};  //! Position in file of cache entry.
+      Int_t fLen{0};  //! Length of cache entry.
+   };
```
Add new line after struct definition.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +590,287 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+      auto offset = iter->fIndex;
+      //printf("Expecting data at offset %ld in miss cache.\n", offset);
+      memcpy(buf, &(fMissCache->fData[offset]), len);
+      fNMissReadOk++;
```
Prefer ++fNMissReadOk unless you need the postfix behavior.

bbockelm commented:
@pcanal - can you re-review?  I believe all requested changes are done.

pcanal commented on tree/tree/src/TTreeCache.cxx:
```diff
@@ -614,10 +586,296 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+
+   if (iter != fMissCache->fEntries.end()) {
+      if (len > iter->fIO.fLen) {
+         fNMissReadMiss++;
```
Could you use prefix?

bbockelm replied:
Hi @pcanal -

Running through open PRs.  The above item is the only requested change, but (perhaps it's just the fact it is late) I can't quite figure out what your comment above means.  Could you elaborate a bit?

Brian

pcanal replied:
I mean 'prefix addition'
```
++fNMissReadMiss;
```
Not an essential change.

pcanal commented:
It looks like this needs a corresponding roottest PR.

bbockelm commented:
@pcanal - corresponding PR for roottest is https://github.com/root-project/roottest/pull/64 .  Unfortunately, I think they need to go in synchronously.

bbockelm commented:
@pcanal - the last comment above has been addressed.

etejedor commented:
@bbockelm could you have a look at the two tree cache tests that are failing for this one? Thanks

pcanal commented:
@phsft-bot build

pcanal commented:
@phsft-bot build
